src/test/assembly/asm/powerpc-types.rs

```diff
@@ -0,0 +1,80 @@
+// min-llvm-version: 10.0.1
+// assembly-output: emit-asm
+// compile-flags: --target powerpc-unknown-linux-gnu
+// needs-llvm-components: powerpc
+
+#![feature(no_core, lang_items, rustc_attrs, repr_simd)]
+#![crate_type = "rlib"]
+#![no_core]
+#![allow(asm_sub_register, non_camel_case_types)]
+
+#[rustc_builtin_macro]
+macro_rules! asm {
+    () => {};
+}
+#[rustc_builtin_macro]
+macro_rules! concat {
+    () => {};
+}
+#[rustc_builtin_macro]
+macro_rules! stringify {
+    () => {};
+}
+
+#[lang = "sized"]
+trait Sized {}
+#[lang = "copy"]
+trait Copy {}
+
+type ptr = *const i32;
+
+impl Copy for i8 {}
+impl Copy for u8 {}
+impl Copy for i16 {}
+impl Copy for i32 {}
+impl Copy for i64 {}
+impl Copy for f64 {}
+impl Copy for ptr {}
+extern "C" {
+    fn extern_func();
+    static extern_static: u8;
+}
+
+// Hack to avoid function merging
+extern "Rust" {
+    fn dont_merge(s: &str);
+}
+
+macro_rules! check { ($func:ident, $ty:ty, $class:ident, $mov:literal) => {
+    #[no_mangle]
+    pub unsafe fn $func(x: $ty) -> $ty {
+        dont_merge(stringify!($func));
+
+        let y;
+        asm!(concat!($mov," {}, {}"), out($class) y, in($class) x);
+        y
+    }
+};}
+
+macro_rules! check_reg { ($func:ident, $ty:ty, $reg:tt, $mov:literal) => {
+    #[no_mangle]
+    pub unsafe fn $func(x: $ty) -> $ty {
+        dont_merge(stringify!($func));
+
+        let y;
+        asm!(concat!($mov, " ", $reg, ", ", $reg), lateout($reg) y, in($reg) x);
+        y
+    }
+};}
+
+// CHECK-LABEL: reg_i32:
+// CHECK: #APP
+// CHECK: mr {{[0-31]+}}, {{[0-31]+}}
+// CHECK: #NO_APP
+check!(reg_i32, i32, reg, "mr");
+
+// CHECK-LABEL: reg_f64:
+// CHECK: #APP
+// CHECK: fmr {{[0-31]+}}, {{[0-31]+}}
+// CHECK: #NO_APP
+check!(reg_f64, f64, freg, "fmr");
```

Comment:
Add tests for explicit register operands, e.g. `r0`, `r18` and `f0`.

Reply:
Hmm - it appears there may be a problem.
It looks like LLVM uses numbers to refer to registers - e.g. `r0` is simply `0` in LLVM inline assembly.
```
error: invalid operand for instruction
  --> C:\Dev\Repositories\rust\src/test\assembly\asm\powerpc-types.rs:65:14
   |
65 |         asm!(concat!($mov, " ", $reg, ", ", $reg), lateout($reg) y, in($reg) x);
   |              ^
   |
note: instantiated into assembly here
  --> <inline asm>:1:5
   |
1  |     mr r18, r18
   |        ^
```

```
error: invalid register `18`: unknown register
  --> C:\Dev\Repositories\rust\src/test\assembly\asm\powerpc-types.rs:65:69
   |
65 |         asm!(concat!($mov, " ", $reg, ", ", $reg), lateout($reg) y, in($reg) x);
   |                                                                     ^^^^^^^^^^
...
86 | check_reg!(reg_r0, i32, "18", "mr");
   | ------------------------------------ in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

So, I removed the `r` prefix from all of the integer registers to try and match LLVM's assembly style.
However, doing that seems to confuse Rust:
```
error: couldn't allocate output register for constraint '{18}'
  --> C:\Dev\Repositories\rust\src/test\assembly\asm\powerpc-types.rs:65:14
   |
65 |         asm!(concat!($mov, " ", $reg, ", ", $reg), lateout($reg) y, in($reg) x);
   |              ^
```

So I'm not too sure how to fix this one. All the other architectures have different prefixes for registers, whether it's `$` or `r` or another variation.

Reply:
No, that's not right: LLVM uses `rX` internally to refer to registers, not `X`.

"invalid operand for instruction" means that the assembler doesn't like `r18` as an operand for `mr`, which seems strange to me.

Reply:
OK it seems that the assembler only accepts `18` as an operand, but LLVM only accepts `r18` for the register name. You can support `"18"` as an alias for `"r18"` in the register definitions.

Reply:
Hmm, only problem is they do the same thing for floating point registers... Is there a way to tell the assembler not to do this?
This is inconsistent with every other inline asm architecture.

Reply:
This is how PPC assembly works, we have no control over it.

Reply:
Okay - this will work for the integer registers, but note that they do the same thing with floating-point registers and Rust doesn't like that:
```
error: unreachable pattern
   --> compiler\rustc_target\src\asm\mod.rs:92:27
    |
55  | /  macro_rules! def_regs {
56  | |      ($arch:ident $arch_reg:ident $arch_regclass:ident {
57  | |          $(
58  | |              $reg:ident: $class:ident $(, $extra_class:ident)* = [$reg_name:literal $(, $alias:literal)*] $(% $filter:ident)?,
...   |
92  | |                          $($alias)|* | $reg_name => {
    | |                            ^^^^^^
...   |
129 | |      }
130 | |  }
    | |__- in this expansion of `def_regs!`
    |
   ::: compiler\rustc_target\src\asm\powerpc.rs:45:1
    |
45  |  / def_regs! {
46  |  |     PowerPC PowerPCInlineAsmReg PowerPCInlineAsmRegClass {
47  |  |         r0: reg = ["r0", "0"],
48  |  |         r3: reg, reg_nonzero = ["r3", "3"],
...    |
123 |  |     }
124 |  | }
    |  |_- in this macro invocation
    |
    = note: `-D unreachable-patterns` implied by `-D warnings`
```

Reply:
The names in `[]` are just aliases that can be used when specifying a register in `in("reg_name")`. They are unrelated to the name that actually appear in the assembly code (which is defined in `emit` and by LLVM).

Reply:
Ah okay - I was going to alias the floating point registers the same way but that's not necessary.
I can just change the `check_reg` macro to take separate register constraint and operand literals.